spec/integration/active_record_spec.rb

```diff
@@ -1,88 +1,120 @@
 # frozen_string_literal: true
 
-require "spec_helper"
 require "fileutils"
 
 RSpec.describe "ActiveRecord failover", type: :active_record do
   EXPECTED_POSTS_COUNT = "100"
 
-  def start_dummy_rails_server
-    raise "Could not start dummy server" if !system("make start_dummy_rails_server")
+  def restart_dummy_rails_server
+    stop_rails_server
+    start_rails_server
   end
 
-  def stop_dummy_rails_server
-    system("make stop_dummy_rails_server")
+  # rubocop:disable RSpec/BeforeAfterAll
+  before(:all) do
+    start_pg_primary
+    start_pg_replica
+    setup_rails_server
+    start_rails_server
   end
 
-  def restart_dummy_rails_server
-    stop_dummy_rails_server
-    start_dummy_rails_server
+  after do
+    start_pg_primary
+    start_rails_server
   end
 
-  # rubocop:disable RSpec/BeforeAfterAll
-  before(:all) { start_dummy_rails_server }
+  after(:all) do
+    stop_rails_server
+    teardown_rails_server
+    stop_pg_replica
+    stop_pg_primary
+  end
+
+  it "should failover to reading connection handler when PG primary is down and fallback to writing connection handler when PG primary is back up" do
+    response = get("/posts")
+
+    expect(response.code.to_i).to eq(200)
+
+    expect(response.body).to eq(<<~BODY.chomp)
+    Posts count: #{EXPECTED_POSTS_COUNT}
+    role: writing
+    BODY
 
-  after(:all) { stop_dummy_rails_server }
+    stop_pg_primary
+
+    get("/posts") # Trigger process to failover
 
-  it "should failover to reading connection handler when PG primary " \
-       "is down and fallback to writing connection handler when PG primary is back up" do
     response = get("/posts")
 
     expect(response.code.to_i).to eq(200)
-    expect(response.body).to include("writing")
-    expect(response.body).to include(EXPECTED_POSTS_COUNT)
 
-    system("make stop_pg_primary")
+    expect(response.body).to eq(<<~BODY.chomp)
+    Posts count: #{EXPECTED_POSTS_COUNT}
+    role: reading
+    BODY
 
-    flood_get("/posts", times: 10) # Trigger all processes to failover
+    start_pg_primary # Start the fallback process
 
-    flood_get("/posts", times: 100) do |res|
-      expect(res.code.to_i).to eq(200)
-      expect(res.body).to include("reading")
-      expect(res.body).to include(EXPECTED_POSTS_COUNT)
-    end
-  ensure
-    system("make restart_pg_primary")
+    sleep 0.05 # Wait for fallback to complete
+
+    response = get("/posts")
+
+    expect(response.body).to eq(<<~BODY.chomp)
+    Posts count: #{EXPECTED_POSTS_COUNT}
+    role: writing
+    BODY
```

Comment:
This was missing which is why our integration test suite did not catch the problem. We were not testing the fallback process.